File: rust/463.island-perimeter.rs
/*
 * @lc app=leetcode id=463 lang=rust
 *
 * [463] Island Perimeter
 */

// @lc code=start
impl Solution {
    pub fn island_perimeter(grid: Vec<Vec<i32>>) -> i32 {
        let mut perimeter = 0;
        let r = grid.len();
        for row in 0..grid.len() {
            for col in 0..grid[0].len() {
                if grid[row][col] == 1 {
                    perimeter += 4;
                    // up
                    match grid.get(row - 1) {
                        Some(r) if r[col] == 1 => {
                            perimeter -= 1;
                        }
                        _ => (),
                    }
                    // down
                    match grid.get(row + 1) {
                        Some(r) if r[col] == 1 => {
                            perimeter -= 1;
                        }
                        _ => (),
                    }
                    // left
                    match grid[row].get(col - 1) {
                        Some(g) if *g == 1 => {
                            perimeter -= 1;
                        }
                        _ => (),
                    }
                    // right
                    match grid[row].get(col + 1) {
                        Some(g) if *g == 1 => {
                            perimeter -= 1;
                        }
                        _ => (),
                    }
                }
            }
        }
        perimeter
    }
}
// @lc code=end

File: javascript/463.island-perimeter.js
/*
 * @lc app=leetcode id=463 lang=javascript
 *
 * [463] Island Perimeter
 */

// @lc code=start
/**
 * @param {number[][]} grid
 * @return {number}
 */
var islandPerimeter = function (grid) {
  let perimeter = 0;
  for (let r = 0; r < grid.length; r++) {
    for (let c = 0; c < grid[r].length; c++) {
      if (grid[r][c] === 1) {
        perimeter += 4;
        if (r > 0 && grid[r - 1][c] === 1) perimeter--;
        if (r < grid.length - 1 && grid[r + 1][c] === 1) perimeter--;
        if (c > 0 && grid[r][c - 1] === 1) perimeter--;
        if (c < grid[r].length - 1 && grid[r][c + 1] === 1) perimeter--;
      }
    }
  }
  return perimeter;
};
// @lc code=end
